db.go

```diff
@@ -309,17 +314,82 @@ func (db *DB) GetMany(ro *ReadOptions, keys [][]byte) ([][]byte, []error) {
 		if valueLens[i] > 0 {
 			values[i] = C.GoBytes(unsafe.Pointer(uintptr(unsafe.Pointer(cPackedVals))+uintptr(valOffset)), C.int(valueLens[i]))
 			valOffset += int(valueLens[i])
-		} else if (errs == nil || errs[i] == nil) && valueLens[i] == 0 {
+		} else if (err == nil || err.(*MultiKeyError).errAt(i) == nil) && valueLens[i] == 0 {
 			// No error and no value, it means keys[i] is found but with empty value
 			values[i] = []byte{}
 		}
 		// else we have either:
-		// 1. (errs != nil && errs[i] == nil) && valueLens[i] < 0 indicating keys[i] is not found
-		// 2. (errs != nil && errs[i] != nil) indicating an error looking up keys[i].
+		// 1. (err != nil && multiKeyErr.errAt(i) == nil) && valueLens[i] < 0 indicating keys[i] is not found
+		// 2. (err != nil && multiKeyErr.errAt(i) != nil) indicating an error looking up keys[i].
 		// In both cases values[i] defaults to nil
 	}
 
-	return values, errs
+	return values, err
+}
+
+// MultiKeyError encapsulates multiple errors encountered in a Get/Put-Many() call,
+// behind the errors.error interface. Caller may cast the generic error object to
+// a MultiKeyError and call Errors() and/or FailedKeyIndexes() to get additional details.
+type MultiKeyError struct {
+	// invariant: errsByIdx is either nil or has at least ONE entry, it's never an empty map
+	errsByIdx map[int]error
+}
+
+func (mke *MultiKeyError) Error() string {
+	if mke.errsByIdx == nil {
+		return ""
+	}
+	return "Multiple keys encountered error, cast via err.(*levigo.MultiKeyErrors).Errors() to get all the errors"
+}
+
+func (mke *MultiKeyError) GoString() string {
+	return fmt.Sprintf("*%#v", *mke)
+}
+
+// FailedKeyIndexes() gives a list of indexes for which Get/PutMany() failed
+// for key at keys[indexes[i]]; e.g. if this returns[2, 3, 7], it means
+// Get/PutMany({keys[2], keys[3], keys[7]}) failed.
+func (mke *MultiKeyError) FailedKeyIndexes() []int {
+	if mke.errsByIdx == nil {
+		return nil
+	}
+
+	indexes := make([]int, 0, len(mke.errsByIdx))
+	for i := range mke.errsByIdx {
+		indexes = append(indexes, i)
+	}
+	return indexes
+}
+
+// Errors() gives specific errors failed for each key from FailedKeyIndexes()
+func (mke *MultiKeyError) Errors() []error {
+	if mke.errsByIdx == nil {
+		return nil
+	}
+
+	errs := make([]error, 0, len(mke.errsByIdx))
+	for i := range mke.errsByIdx {
+		errs = append(errs, mke.errsByIdx[i])
+	}
+	return errs
+}
+
+func (mke *MultiKeyError) addKeyErr(i int, err error) {
+	if mke.errsByIdx == nil {
+		mke.errsByIdx = make(map[int]error)
+	}
+	mke.errsByIdx[i] = err
+}
+
+func (mke *MultiKeyError) errAt(i int) error {
```

Comment:
Should this be a public method ? `ErrAt()` since `FailedKeyIndexes()` is public

Reply:
for now I am only letting user do:
```
      if mke, ok := err.(*levigo.MultiKeyError); ok {
          mke.FailedKeyIndexes(), mke.Errors() := errs, indexes
          for i := range errs {
              fmt.Printf("Failed for key %d, error: %s", indexes[i], errs[i])
          }
      }
```
you only get to know for what key index the get failed. This is probably safer.